### ISSUE TITLE
Scroll wheel scrolling is not smooth

### ISSUE THREAD

**Issue by ds84182** [flutter/flutter#32120]
## Steps to Reproduce

1. Use a desktop embedder, or another platform that allows you to scroll with a mouse's scroll wheel.
2. Scroll any scrollable content (e.g. list demo in Flutter Gallery)
3. Observe how the scroll offset abruptly jumps between scroll notches instead of smoothly transitioning (like Chrome, Firefox, UWP apps, etc.)

Gif example:
![2019-05-05_11-10-31](https://user-images.githubusercontent.com/2268005/57196013-9756cf80-6f26-11e9-8216-e2399fea16f3.gif)

**Comment by ds84182** [comment]
I ported Chrome's rather complex smooth scrolling algorithm (see https://bit.ly/smoothscrolling)

![2019-05-06_10-23-09](https://user-images.githubusercontent.com/2268005/57231731-1b718b80-6fe9-11e9-976e-08347d604ecd.gif)

I can clean it up a bit (make the code a bit more Dart-y) and make a PR, if using Chrome's smooth scrolling is acceptable.

**Comment by stuartmorgan** [comment]
/cc @gspencergoog 

It seems like this is going to need a design proposal, or at least discussion, for how it would be controlled, since it should almost certainly be optional. Smooth scrolling with scroll wheels is definitely not universal behavior on desktop platforms.

**Comment by ds84182** [comment]
Notably, Windows (UWP apps) and Mac OS have their own smooth scrolling behaviors. Linux doesn't really have a default, so I feel that using Chrome's implementation (at least on Linux) would be good enough.

**Comment by ds84182** [comment]
For UWP, a lot of the logic can be found [here (Scroller.cpp)](https://github.com/Microsoft/microsoft-ui-xaml/blob/ca0ad2b4a7b09527dba7a139c8dd3f6300aace9a/dev/Scroller/Scroller.cpp), and some keyboard scrolling logic can be found [here (ScrollViewer.cpp)](https://github.com/Microsoft/microsoft-ui-xaml/blob/ca0ad2b4a7b09527dba7a139c8dd3f6300aace9a/dev/ScrollViewer/ScrollViewer.cpp#L1899)

**Comment by gspencergoog** [comment]
Marking this as a `framework` issue, since I don't think it's specific to any one platform, and the solution is probably in the framework and not in the engine.

**Comment by stuz5000** [comment]
Is anyone working on this? I see this problem in the Android emulator. Using the scrollwheel leads to very erratic behavior. Sometimes it scrolls the list down, them pops back.

**Comment by BendixMa** [comment]
@ds84182 Your smooth scroll algorithm looks pretty good from what I can see in the GIF. Could you share your code on how you implemented it? I have flutter web project and a smooth scroll algorithm would greatly increase the user experience. As far as I can tell, every browser supports smooth scrolling and so a flutter website feels a bit odd in comparison to other native websites when being scrolled.

**Comment by ASSASSIN-363** [comment]
> @ds84182 Your smooth scroll algorithm looks pretty good from what I can see in the GIF. Could you share your code on how you implemented it? I have flutter web project and a smooth scroll algorithm would greatly increase the user experience. As far as I can tell, every browser supports smooth scrolling and so a flutter website feels a bit odd in comparison to other native websites when being scrolled.

Do u have the smooth scroll algorithm can u share it ?

**Comment by markusaksli-nc** [comment]
Reproducible on the latest master `1.21.0-6.0.pre.160`.
<details>
 <summary>Minimal reproducible code sample</summary>

 ```
import 'dart:math';

import 'package:flutter/material.dart';

void main() => runApp(
      MaterialApp(
        home: Scaffold(
          appBar: AppBar(
            title: Text("Sample"),
          ),
          body: ListView(
            children: [
              ...List.generate(200, (index) => RandomBox()),
            ],
          ),
        ),
      ),
    );

class RandomBox extends StatelessWidget {
  static const List colors = [
    Colors.red,
    Colors.green,
    Colors.yellow,
    Colors.blue,
    Colors.purple,
    Colors.black,
    Colors.orange
  ];
  static Random random = new Random();

  @override
  Widget build(BuildContext context) {
    int index = random.nextInt(7);
    return Container(
      height: 50,
      color: colors[index],
    );
  }
}
```
</details>
<details>
 <summary>flutter doctor -v</summary>

 ```
 [✓] Flutter (Channel master, 1.21.0-6.0.pre.160, on Mac OS X 10.15.6 19G73, locale en-GB)
     • Flutter version 1.21.0-6.0.pre.160 at /Users/markus/development/flutter_master
     • Framework revision 31ee51a302 (11 hours ago), 2020-08-03 17:27:21 -0700
     • Engine revision fa16c3d0ba
     • Dart version 2.10.0 (build 2.10.0-2.0.dev 0f0e04ec3a)

  
 [✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
     • Android SDK at /Users/markus/Library/Android/sdk
     • Platform android-30, build-tools 30.0.1
     • Java binary at: /Users/markus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/193.6626763/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
     • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
     • All Android licenses accepted.

 [✓] Xcode - develop for iOS and macOS (Xcode 11.6)
     • Xcode at /Applications/Xcode.app/Contents/Developer
     • Xcode 11.6, Build version 11E708
     • CocoaPods version 1.9.3

 [✓] Chrome - develop for the web
     • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

 [✓] Android Studio (version 4.0)
     • Android Studio at /Users/markus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/193.6626763/Android Studio.app/Contents
     • Flutter plugin version 48.0.2
     • Dart plugin version 193.7361
     • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

 [✓] Connected device (5 available)
     • sdk gphone x86 arm (mobile) • emulator-5554                            • android-x86    • Android 11 (API 30) (emulator)
     • Nevercode’s iPhone (mobile) • b668e524315069f3db3661ac11ff1f66afafebdb • ios            • iOS 13.6
     • macOS (desktop)             • macos                                    • darwin-x64     • Mac OS X 10.15.6 19G73
     • Web Server (web)            • web-server                               • web-javascript • Flutter Tools
     • Chrome (web)                • chrome                                   • web-javascript • Google Chrome 84.0.4147.105

 • No issues found!
```
</details>

**Comment by iapicca** [comment]
a partial, non-resolutive [workaround](https://github.com/flutter/flutter/issues/35687#issuecomment-685789907),
an improvement nonetheless

**Comment by nonoyona** [comment]
Can't wait until December!

**Comment by Piinks** [comment]
@markusaksli-nc in your reproduction of this issue, what device were you executing that sample code on? I have not been able to identify the jank here.

**Comment by stuartmorgan** [comment]
@Piinks This isn't about jank per se, but about smooth-scroll behavior when using scroll wheels.

**Comment by Piinks** [comment]
Hey @stuartmorgan! Thanks! 
I've been trying to understand what the expectation here is. There are mentions of web and desktop above, but in discussing with @chinmaygarde it appeared this is WAI for desktop as it matches native behavior. Is this web specific?

**Comment by Piinks** [comment]
I've been playing with [this flutter web app](https://flutterwebsitetest1.web.app/#/) to check scrolling behavior on web and found it behaves differently with different types of mouse scroll wheels.
For example, it is smooth using an Apple magic mouse, but when using an incremental/click scroll wheel, it does not move as smoothly. Is the expectation that an incremental scroll wheel will simulate a smooth animation? I am wondering if the device should dictate the behavior as it is doing now.

**Comment by stuartmorgan** [comment]
> in discussing with @chinmaygarde it appeared this is WAI for desktop as it matches native behavior. Is this web specific?

It depends on the platform; comments above mentioned that UWP uses smooth scrolling, for instance (I'm not familiar enough with UWP to know if that's always true, or an optional behavior). Even for web it is (or at least was; I haven't compared recently) platform-specific on at least some browsers. IIRC when it was implemented in Chrome it was turned on for Windows and Linux but not macOS, for instance.

> Is the expectation that an incremental scroll wheel will simulate a smooth animation?

Yes, smooth-scroll is specifically about synthesizing a smooth scroll animation in the face of discrete, large deltas (like most scroll wheels use)

> I am wondering if the device should dictate the behavior as it is doing now.

We could certainly decide not to support smooth-scroll, but having it implemented with the ability to enable or disable it to match the platform expectations is, IMO, a reasonable feature request.

**Comment by Piinks** [comment]
Ah! Thank you!
So we're looking for

> a smooth scroll animation in the face of discrete, large deltas

that can be turned on/off at the user's discretion. SGTM! I think that is reasonable as well! I think I understand now, let me know if I am still off the mark. I'll update the title as well.

**Comment by iapicca** [comment]
@Piinks 
I feel that  the previous title `Scroll wheel scrolling is not smooth` implies an **issue** 
while `Add option to smoothly animate large mouse scroll deltas` seems more like a feature request

Mouse wheel scrolling experience may be somewhat unrefined in the sample [above](https://github.com/flutter/flutter/issues/32120#issue-440455298)
but using a [PageView](https://api.flutter.dev/flutter/widgets/PageView-class.html) I'm sorry to say that is a poor experience

Here's a [code sample](https://dartpad.dartlang.org/3eecda9104056871e7e887aaf765900e), a [video](https://photos.app.goo.gl/GtwV7MsH82rCbqkG7) and a possible  [workaround](https://stackoverflow.com/a/63706366/9836706)
in discord chat I noticed that I'm not the only one experiencing this situation ([link](https://discordapp.com/channels/608014603317936148/608016344520196173/747844032813858856))

Nobody can complain about adding a more granular option,
but imho it should added on top of a solid base experience

lastly I'd like to reference this [issue](https://github.com/flutter/flutter/issues/35687) which is specific for the case I described,
but with lower priority that the present

**Comment by Piinks** [comment]
Hi @iapicca!

I am not sure I understand, if the issue you mentioned is covered by #35687, I don't know that we need to include it here. I am looking to make this bug specific and actionable so we can resolve it. `Scroll wheel scrolling is not smooth` is rather vague. 😜  

As I understand it, this issue is a feature request. Current mouse scrolling behavior matches native platforms, and so this appears to be a request for a generic smoothing feature that users can opt into (paraphrased from @chinmaygarde 🙂 ).

**Comment by iapicca** [comment]
@Piinks I'm just saying that it would be great if this new feature would work with [PageView](https://api.flutter.dev/flutter/widgets/PageView-class.html) as well

**Comment by Piinks** [comment]
After conferring with @gspencergoog & @chinmaygarde, this will require a generic smoothing algorithm and the ability to allow users to opt into using it. Matching a specific browsers implementation should to be out of scope.
Since we are currently matching native scroll behavior, I am un-assigning myself to work on resolving scrolling bugs like #35687. :)

**Comment by ImDeZz** [comment]
Here is a guide showing how to make your own scroll animation that's fluid.
https://hobbister.com/2020/11/17/smooth-scrolling-with-flutter-web/

From this:
![gif0](https://user-images.githubusercontent.com/18382875/99574281-74d8ba00-29d7-11eb-9d0a-a6dd9eee7d9b.gif)

To this:
![gif1](https://user-images.githubusercontent.com/18382875/99574308-80c47c00-29d7-11eb-8aa7-8d535e322b72.gif)

**Comment by zibizz1** [comment]
Mouse scroll is not working when the cursor is on multiline text
```
TextField(
    keyboardType: TextInputType.multiline,
    minLines: 5,
    maxLines: null,
    );
```
Scrolling stops when text field hits the cursor during scrolling. It's not so great.

**Comment by Piinks** [comment]
@zibizz1 that sounds like a separate issue. Can you file a separate issue with a reproducible code sample and the output of `flutter doctor -v`?

**Comment by lazylazyllama** [comment]
> After conferring with @gspencergoog & @chinmaygarde, this will require a generic smoothing algorithm and the ability to allow users to opt into using it. Matching a specific browsers implementation should be out of scope.

As any modern browser has smooth interpolated scrolling enabled by default, o think flutter should enable the smooth scrolling algorithm for web by default. Same goes for uwp. I don't know about other platforms, but in general I think that all are going towards smooth scrolling, so just enabling it for every platform should also be considered.

When ever I open a flutter web app on my browser on windows it really feels extremely weird not to have the mouse scroll animated :(

This really is an issue that stops me from using flutter for my web projects right now, which would be so cool to do :(

**Comment by Wissperwind** [comment]
Any updates on this?
I am also looking for an option to achieve smoothe scrolling with the standard scrollbar in flutter web. If it is a default of a parameter is not that important to me. I could live quite wenn with Scrollbar(smoothScrolling: true, ...).

**Comment by gaaclarke** [comment]
I think designating this as a "Feature Request" belies the severity of the issue.  This is one of those things that developers don't take seriously but designers/users lose their minds.  Other places where we don't have parity with the platform we are targeting, we treat it like a bug and not a "feature request".  This could even cause sensitive users nausea depending on the circumstances (I'm thinking things like animations tied to scrolling offset).

**Comment by iapicca** [comment]
> I think designating this as a "Feature Request" belies the severity of the issue. This is one of those things that developers don't take seriously but designers/users lose their minds. Other places where we don't have parity with the platform we are targeting, we treat it like a bug and not a "feature request". This could even cause sensitive users nausea depending on the circumstances (I'm thinking things like animations tied to scrolling offset).

@gaaclarke I agree, I'd briefly add that this issue affect **_heavily_** the behavior of the `pageview` as pointed out [here](https://github.com/flutter/flutter/issues/35687#issuecomment-639273756)

> ![image](https://user-images.githubusercontent.com/45637986/112061530-49253680-8b67-11eb-8c37-c0a9299f838f.png)

it does **_feel_** like a bug

**Comment by yjbanov** [comment]
I think not only smooth scrolling should be an option, but it should be the default. I strongly agree with https://github.com/flutter/flutter/issues/32120#issuecomment-804380127. I think this should be P3 at least.

**Comment by franarolas** [comment]
Any news on this? this issue has been out for more than 2 years now...

**Comment by iapicca** [comment]
@TahaTesser  could you consider the label `crowd` for this issue?

**Comment by CiriousJoker** [comment]
Any news?

I've used @ImDeZz workaround, but its quite hacky, especially since I needed to modify it to make it compatible with touch devices by switching it on/off depending on the device.

**Comment by yjbanov** [comment]
I'm going to raise the priority of this to P3 as this is very noticeable on the web, where all pages/apps except Flutter Web scroll smoothly.

The full analysis of scroll animations will probably take some time (the browser-OS matrix is not small). So I propose that we fix this in multiple phases. The first phase would be a "simple fix", as follows:

- We choose one reasonable default scroll animation and use it on all browsers on all operating systems. For example, the [Chromium animation](https://bit.ly/smoothscrolling) that @ds84182 pointed to in https://github.com/flutter/flutter/issues/32120#issuecomment-489640098 should already improve things significantly.
- We make the animation customizable, so that the developer community can implement their own animations and/or create pub packages to share with everyone.

After this phase we continue listening to developer feedback and decide what further steps are necessary. It is possible that the reasonable default and the package ecosystem together fix the problem enough that no further work is necessary (some developers are coming up with really interesting ideas, such as [this](https://github.com/flutter/flutter/issues/77883) and [this](https://github.com/flutter/flutter/issues/82406)). If not, we'll get more specific feedback and address it as it arrives.

**Comment by SilverMarcs** [comment]
Weren't they saying something about this in the 2.10 Windows update video? Something about smooth scrolling for windows desktop apps at least.

**Comment by iapicca** [comment]
@Piinks @yjbanov 
please consider adding the issue to [What's Next on Web Scroll Behaviors](https://github.com/flutter/flutter/projects/224)

**Comment by Piinks** [comment]
@iapicca as stated in the project description and in the former tracking issue (#90559), that scope of work is based on [a published article](https://medium.com/flutter/flutter-whats-next-on-the-web-e0454bff964) and intended for tracking the progress of work on the stated commitment. 

Soliciting and campaigning for issues on other issues does nothing to progress them towards resolution.

This issue is prioritized and labeled in accordance with our issue hygiene. We are committed to fixing it. Please read the [issue hygiene](https://github.com/flutter/flutter/wiki/Issue-hygiene) section of our wiki if you are unfamiliar.

**Comment by alexmercerind** [comment]
> Here is a guide showing how to make your own scroll animation that's fluid. https://hobbister.com/2020/11/17/smooth-scrolling-with-flutter-web/

@ImDeZz, thanks for your package.


It's scroll behaviour seems to be good with mouse scroll-wheel.
However, I think it's not very good with touchpads (too fast and jittery).

Apparently, apps seem to scroll with a _different behaviour_ when touchpad is used (instead of mouse scroll-wheel). This scrolling is very similar to how touch-screen scroll behaves.
Same can be seen in UWP apps & Chromium browsers (Microsoft Edge still seems to have different smooth scrolling algorithm compared to Google Chrome).

With that being said, there doesn't seem to be a way to distinguish between a "touchpad scroll" & "mouse scroll-wheel scroll" yet.
- https://github.com/flutter/flutter/issues/23604
- https://github.com/flutter/flutter/issues/88175

So, I think touch-pad smooth scrolling should also be taken in account (since it currently follows same mouse scroll events).

Thankyou.
Looking forward.

**Comment by omaru-ok** [comment]
yes please sir, i want smooth scrooll  :( 

pp please

**Comment by exaby73** [comment]
While trackpad scrolling is smooth, scrolling with a mouse wheel is not. Reproduced on `Master (3.10.0-10.0.pre.39)`.

<details><summary><code>flutter doctor -v</code> (Master)</summary>

```
[!] Flutter (Channel master, 3.10.0-10.0.pre.39, on macOS 13.3.1 22E261 darwin-arm64, locale en-US)
    • Flutter version 3.10.0-10.0.pre.39 on channel master at /Users/nabeelparkar/fvm/versions/master
    ! Warning: `dart` on your path resolves to /opt/homebrew/Cellar/dart/2.19.5/libexec/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nabeelparkar/fvm/versions/master. Consider adding /Users/nabeelparkar/fvm/versions/master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d186792c00 (3 hours ago), 2023-04-20 20:50:59 -0700
    • Engine revision 122c3b3820
    • Dart version 3.1.0 (build 3.1.0-26.0.dev)
    • DevTools version 2.23.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at /Users/nabeelparkar/Library/Android/sdk/
    • Platform android-33, build-tools 33.0.2
    • ANDROID_SDK_ROOT = /Users/nabeelparkar/Library/Android/sdk/
    • Java binary at: /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.0

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Brave Browser.app/Contents/MacOS/Brave Browser

[✓] Android Studio (version 2022.1)
    • Android Studio at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/221.6008.13.2211.9619390/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.77.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.62.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.3.1 22E261 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Brave Browser 112.1.50.119

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>

**Comment by shroot91** [comment]
> While trackpad scrolling is smooth, scrolling with a mouse wheel is not. Reproduced on `Master (3.10.0-10.0.pre.39)`.
> `flutter doctor -v` (Master)

Neither on the web app D:

**Comment by DexEze** [comment]
This is still an issue with mouse on all browsers. please fix this

**Comment by iapicca** [comment]
@exaby73
I think @DexEze has a point, this issue affect desktop as much as web

**Comment by water-mizuu** [comment]
Any updates on this?

**Comment by ZiyadF296** [comment]
Any updates on this?

**Comment by water-mizuu** [comment]
It's almost been 4 months since the last change here, any updates?

**Comment by DexEze** [comment]
Comment on https://github.com/flutter/flutter/pull/115314 and tag @Piinks

**Comment by ax-null** [comment]
Any updates?

**Comment by iapicca** [comment]
@Piinks 
is [this](https://github.com/flutter/flutter/issues/32120#issuecomment-725913608) the workaround related to the tag
or there is something better out there

**Comment by DexEze** [comment]
@Piinks can you please clarify what the workaround is for this?

**Comment by asagarshinde** [comment]
@Piinks please update, I have created a complete web app using Flutter and now while scrolling it does not seem natural.

**Comment by knottx** [comment]
@Piinks hope you are doing well.

**Comment by clragon** [comment]
I looked at the available workarounds and saw [`smooth_scroll_multiplatform`](https://pub.dev/packages/smooth_scroll_multiplatform) and [`dyn_mouse_scroll`](https://pub.dev/packages/dyn_mouse_scroll).

However, they depend on Provider for no apparent reason, dont allow specifying a ScrollController despite not needing to create their own internally and dont respect the default ScrollPhysics of the surrounding Platform on Mobile devices.

To rectify these things I created a slightly less janky version below. I will not make this a package and the original repositories dont seem maintained well so I will not make a PR either. But feel free to directly use this code in your project. 

Do note that this and all "workarounds" probably suffer from the same problem that is currently blocking the resolution of this issue. Namely, trackpads might break when using this: https://github.com/flutter/flutter/issues/139990

Additionally please stop tagging the developers. 
This does nothing to make this go faster but instead notifies everyone in this thread unnecessarily.
See here: https://github.com/flutter/flutter/wiki/Issue-hygiene#when-will-my-bug-be-fixed

Workaround code:

<details open><summary>scroll_animator.dart</summary>

```dart
import 'dart:math' as math;

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

class ScrollAnimator extends StatefulWidget {
  /// Wraps a scrollable widget to animate scroll delta on desktop.
  const ScrollAnimator({
    super.key,
    this.controller,
    this.duration = const Duration(milliseconds: 380),
    this.scrollSpeed = 2,
    this.curve = Curves.easeOutQuart,
    required this.builder,
  });

  /// The scroll controller to use.
  ///
  /// If not provided, a new controller will be created.
  final ScrollController? controller;

  /// The duration of the scroll animation.
  final Duration duration;

  /// The speed at which we scroll.
  final double scrollSpeed;

  /// The curve of the scroll animation.
  final Curve curve;

  /// Builder function for the inner scrollable widget.
  final Widget Function(
    BuildContext context,
    ScrollController controller,
    ScrollPhysics? physics,
  ) builder;

  @override
  State<ScrollAnimator> createState() => _ScrollAnimatorState();
}

class _ScrollAnimatorState extends State<ScrollAnimator> {
  late _ScrollAnimatorController scrollState = _ScrollAnimatorController(
    controller: widget.controller,
    duration: widget.duration,
  );

  @override
  Widget build(BuildContext context) {
    return ListenableBuilder(
      listenable: scrollState,
      builder: (context, _) {
        scrollState.handlePipelinedScroll();
        return Listener(
          onPointerSignal: (signalEvent) => scrollState.handleDesktopScroll(
            signalEvent,
            widget.scrollSpeed,
            widget.curve,
          ),
          onPointerDown: scrollState.handleTouchScroll,
          child: widget.builder(
            context,
            scrollState.controller,
            scrollState.physics,
          ),
        );
      },
    );
  }
}

class _ScrollAnimatorController with ChangeNotifier {
  _ScrollAnimatorController({
    ScrollController? controller,
    this.duration = const Duration(milliseconds: 380),
  }) : controller = controller ?? ScrollController();

  final ScrollController controller;
  final Duration duration;
  ScrollPhysics? get physics => _physics;
  ScrollPhysics? _physics;

  void handlePipelinedScroll() async {
    await Future.value();
    _handlePipelinedScroll?.call();
  }

  VoidCallback? _handlePipelinedScroll;

  double _futurePosition = 0;

  bool _prevDeltaPositive = false;
  double? _lastLock;

  Future<void>? _animationEnd;

  static double _calcMaxDelta(ScrollController controller, double delta) {
    return delta > 0
        ? math.min(controller.position.pixels + delta,
                controller.position.maxScrollExtent) -
            controller.position.pixels
        : math.max(controller.position.pixels + delta,
                controller.position.minScrollExtent) -
            controller.position.pixels;
  }

  void handleDesktopScroll(
    PointerSignalEvent event,
    double scrollSpeed,
    Curve animationCurve, [
    bool readLastDirection = true,
  ]) {
    // Ensure desktop physics is being used.
    if (physics == null || _lastLock != null) {
      if (event is PointerScrollEvent) {
        double posPixels = controller.position.pixels;
        if ((posPixels == controller.position.minScrollExtent &&
                event.scrollDelta.dy < 0) ||
            (posPixels == controller.position.maxScrollExtent &&
                event.scrollDelta.dy > 0)) {
          return;
        } else {
          _physics = const NeverScrollableScrollPhysics();
        }
        bool outOfBounds = posPixels < controller.position.minScrollExtent ||
            posPixels > controller.position.maxScrollExtent;
        double calcDelta = _calcMaxDelta(controller, event.scrollDelta.dy);
        if (!outOfBounds) {
          controller.jumpTo(_lastLock ?? (posPixels - calcDelta));
        }
        double deltaDelta = calcDelta - event.scrollDelta.dy;
        _handlePipelinedScroll = () {
          _handlePipelinedScroll = null;
          double currPos = controller.position.pixels;
          double currDelta = event.scrollDelta.dy;
          bool shouldLock = _lastLock != null
              ? (_lastLock == currPos)
              : (posPixels != currPos + deltaDelta &&
                  (currPos != controller.position.maxScrollExtent ||
                      currDelta < 0) &&
                  (currPos != controller.position.minScrollExtent ||
                      currDelta > 0));
          if (!outOfBounds && shouldLock) {
            controller.jumpTo(posPixels);
            _lastLock = posPixels;
            controller.position.moveTo(posPixels).whenComplete(() {
              _physics = null;
              notifyListeners();
            });
            return;
          } else {
            if (_lastLock != null || outOfBounds) {
              controller.jumpTo(_lastLock != null
                  ? posPixels
                  : (currPos - _calcMaxDelta(controller, currDelta)));
            }
            _lastLock = null;
            handleDesktopScroll(event, scrollSpeed, animationCurve, false);
          }
        };
        notifyListeners();
      }
      return;
    } else if (event is PointerScrollEvent) {
      bool currentDeltaPositive = event.scrollDelta.dy > 0;
      if (readLastDirection && currentDeltaPositive == _prevDeltaPositive) {
        _futurePosition += event.scrollDelta.dy * scrollSpeed;
      } else {
        _futurePosition =
            controller.position.pixels + event.scrollDelta.dy * scrollSpeed;
      }
      _prevDeltaPositive = event.scrollDelta.dy > 0;

      Future<void> animationEnd = _animationEnd = controller.animateTo(
        _futurePosition,
        duration: duration,
        curve: animationCurve,
      );

      animationEnd.whenComplete(() {
        if (animationEnd != _animationEnd) return;
        if (physics == null) return;
        _physics = null;
        notifyListeners();
      });
    }
  }

  void handleTouchScroll(PointerDownEvent event) {
    if (physics == null) return;
    _physics = null;
    notifyListeners();
  }
}
```

</details>

Usage for this snippet is straight forward:

```dart
ScrollAnimator(
  controller: scrollController, // or null
  builder: (context, controller, phyiscs) => ListView(
    controller: controller,
    physics: phyiscs,
  ),
);
```

**Comment by iapicca** [comment]
> I looked at the available workarounds and saw [`smooth_scroll_multiplatform`](https://pub.dev/packages/smooth_scroll_multiplatform) and [`dyn_mouse_scroll`](https://pub.dev/packages/dyn_mouse_scroll).

@clragon 
the package situation seems very chaotic to me

[smooth_scroll_multiplatform](https://pub.dev/packages/smooth_scroll_multiplatform) 
 --> is [fork](https://github.com/alesimula/dyn_mouse_scroll) of @Wizzel1's [repo](https://github.com/Wizzel1/dyn_mouse_scroll) 
 ----> which is a fork of @Bluebar1's [repo](https://github.com/Bluebar1/dyn_mouse_scroll) 
 ------> which is also published as [dyn_mouse_scroll](https://pub.dev/packages/dyn_mouse_scroll)

that said, 
the workaround you mention seem specific to a [ListView](https://api.flutter.dev/flutter/widgets/ListView-class.html), but how about a [PageView](https://api.flutter.dev/flutter/widgets/PageView-class.html)?
Is this [workaround](https://github.com/flutter/flutter/issues/32120#issuecomment-725913608) the best choice for it?

I think it's fair to tag a team member to ask which workaround are they referring to

**Comment by clragon** [comment]
Hi @iapicca,

> the workaround you mention seem specific to a [ListView](https://api.flutter.dev/flutter/widgets/ListView-class.html), but how about a [PageView](https://api.flutter.dev/flutter/widgets/PageView-class.html)?

The workaround I posted seems to work fine with a `PageView`:

```dart
body: ScrollAnimator(
  controller: _controller, // Your PageController
  builder: (context, controller, physics) => PageView(
    scrollDirection: Axis.vertical,
    controller: controller as PageController,
    physics: physics,
  ),
),
```

Its slightly unfortunate that we need to cast here, but you could rewrite the code to use a `T extends ScrollController`.

> I think it's fair to https://github.com/flutter/flutter/issues/32120#issuecomment-1858760397 to team member which workaround are they referring to

The workaround they refer to is listed in the top issue comment:

> There are existing workarounds and packages available
https://pub.dev/packages?q=smooth+scroll

Its simply all packages with a name similar to "smooth scroll".

We can see this as Piinks edited the comment and added that paragraph.
I hope that helps. Best regards.

**Comment by ReinisSprogis** [comment]
I am trying something like this. It is very simple way, but the difference in mouse wheel scroll is very noticeable.  But it doesn't work with touch scroll anymore because I have to provide NeverScrollableScrollPhysics() if I provide AlwaysScrollableScrollPhysics() then it jumps. Ideally could just provide something like SmoothScrollableScrollPhysics() and done. 
```dart
 import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MaterialApp(
    home: Home(),
  ));
}

class Home extends StatefulWidget {
  const Home({super.key});

  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  final ScrollController _scrollController = ScrollController();
  int scrollStartTime = 0;
  int previousScrollTime = 0;
  void _animateTo(double targetPosition) {
    // Ensure any ongoing animation is stopped before starting a new one.
    if (_scrollController.hasClients) {
      int deltaScrollTime = DateTime.now().millisecondsSinceEpoch - previousScrollTime;
      int animationDuration = deltaScrollTime < 500 ? deltaScrollTime : 500;
      _scrollController.animateTo(
        targetPosition,
        duration: Duration(milliseconds: animationDuration),
        curve: Curves.easeOut,
      );
      previousScrollTime = DateTime.now().millisecondsSinceEpoch;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Listener(
        onPointerSignal: (pointerSignal) {
          if (pointerSignal is PointerScrollEvent) {
            final nextPosition = _scrollController.position.pixels + pointerSignal.scrollDelta.dy;
            scrollStartTime = DateTime.now().millisecondsSinceEpoch;
            _animateTo(
              nextPosition,
            );
          }
          return;
        },
        child: ListView(
          controller: _scrollController,
          physics: NeverScrollableScrollPhysics(),
          children: List.generate(500, (index) {
            return ListTile(
              title: Text('Tile index $index'),
            );
          }),
        ),
      ),
    );
  }
}
```

**Comment by shroot91** [comment]
any updates for this issue?

**Comment by oleksii-pimenov** [comment]
With [clragon](https://github.com/clragon) solution mouse wheel scroll work nice but touch scroll become unusable unfortunately

**Comment by oleksii-pimenov** [comment]
Any ideas on how to keep both of them working

**Comment by Mikephii** [comment]
you can override void pointerScroll() on ScrollPosition to use animateTo() instead of forcePixels()
and track the interval between animations the same as clragons solution above

pointerScroll appears to only be called to handle PointerScrollEvent's which are scroll wheel events 
![CleanShot 2024-08-28 at 17 33 01@2x](https://github.com/user-attachments/assets/d1ffdcfc-ee67-4f63-bda4-32b3595a8984)

you can create a customScrollPosition and CustomScrollController to use your custom scroll position. 

This allows you to use any scroll physics and also touch scroll and drag

working great for me not sure of the consequences of my actions but thats usual

**Comment by iapicca** [comment]
> you can override void pointerScroll() on ScrollPosition to use animateTo() instead of forcePixels() and track the interval between animations the same as clragons solution above
> 
> pointerScroll appears to only be called to handle PointerScrollEvent's which are scroll wheel events 
> [...]
> you can create a customScrollPosition and CustomScrollController to use your custom scroll position.
> 
> This allows you to use any scroll physics and also touch scroll and drag
> 
> working great for me not sure of the consequences of my actions but thats usual

@Mikephii 
sounds interesting, mind to share a code sample of your workaround? I'd like to test it

**Comment by oelburk** [comment]
A bit salty here but why should we bother with a workaround on a thing that should just work? This issue has been live since 2019. Any official update from the flutter team regarding this?

**Comment by Mikephii** [comment]
heres a minimal example of my solution. I think this can be used by the flutter team and can easily be a boolean flag in the scroll controller. Unsure of any side effects but it works fine for me and doesnt break 
touch scroll or drag or any of that. 

```import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'dart:math' as math;

class ExampleSmoothScroll extends StatefulWidget {
  const ExampleSmoothScroll({
    super.key,
  });

  @override
  State<ExampleSmoothScroll> createState() => _ExampleSmoothScrollState();
}

class _ExampleSmoothScrollState extends State<ExampleSmoothScroll> {
  final ScrollController _scrollController = MyCustomSmoothScrollController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: ListView.builder(
        physics: BouncingScrollPhysics(),
        controller: _scrollController,
        itemCount: 1000,
        itemBuilder: (context, index) {
          return ListTile(
            tileColor: index % 2 == 0 ? Colors.blue : Colors.lightBlue,
            title: Text('Item $index'),
          );
        },
      ),
    );
  }
}

class MyCustomSmoothScrollController extends ScrollController {
  MyCustomSmoothScrollController({
    double initialScrollOffset = 0.0,
    bool keepScrollOffset = true,
    String? debugLabel,
  }) : super(
          initialScrollOffset: initialScrollOffset,
          keepScrollOffset: keepScrollOffset,
          debugLabel: debugLabel,
        );

  @override
  ScrollPosition createScrollPosition(
    ScrollPhysics physics,
    ScrollContext context,
    ScrollPosition? oldPosition,
  ) {
    //use the customscroll position with smooth scrolling
    return CustomSmoothScrollPosition(
      physics: physics,
      context: context,
      initialPixels: initialScrollOffset,
      keepScrollOffset: keepScrollOffset,
      oldPosition: oldPosition,
      debugLabel: debugLabel,
    );
  }
}

class CustomSmoothScrollPosition extends ScrollPositionWithSingleContext {
  int previousScrollTime = DateTime.now()
      .millisecondsSinceEpoch; //used to track time between scrolls same as clragons code

  CustomSmoothScrollPosition({
    required ScrollPhysics physics,
    required ScrollContext context,
    double initialPixels = 0.0,
    bool keepScrollOffset = true,
    ScrollPosition? oldPosition,
    String? debugLabel,
  }) : super(
          physics: physics,
          context: context,
          initialPixels: initialPixels,
          keepScrollOffset: keepScrollOffset,
          oldPosition: oldPosition,
          debugLabel: debugLabel,
        );

  @override
  void pointerScroll(double delta) {
    // If an update is made to pointer scrolling here, consider if the same
    // (or similar) change should be made in
    // _NestedScrollCoordinator.pointerScroll.
    if (delta == 0.0) {
      goBallistic(0.0);
      return;
    }

    final double targetPixels =
        math.min(math.max(pixels + delta, minScrollExtent), maxScrollExtent);
    if (targetPixels != pixels) {
      goIdle();
      updateUserScrollDirection(
        -delta > 0.0 ? ScrollDirection.forward : ScrollDirection.reverse,
      );
      final double oldPixels = pixels;
      // Set the notifier before calling force pixels.
      // This is set to false again after going ballistic below.
      isScrollingNotifier.value = true;

      //--------------------  NEW CODE STARTS HERE  --------------------
      int deltaScrollTime =
          DateTime.now().millisecondsSinceEpoch - previousScrollTime;
      int animationDuration = deltaScrollTime < 500 ? deltaScrollTime : 500;

      if (animationDuration < 32) {
        //60fps is 16ms per frame so less than 32ms is pointless as its less than 2 frames
        //also avoids calling animateTo too many times in quick succession which seems to be buggy and throws exceptions for elapsed<0  that might be related to github.com/flutter/flutter/issues/106277
        super.pointerScroll(delta);
        return;
      }
      animateTo(targetPixels,
          duration: Duration(milliseconds: animationDuration),
          curve: Curves.linear);
      previousScrollTime = DateTime.now().millisecondsSinceEpoch;
      //--------------------  NEW CODE ENDS HERE  --------------------

      //-------------------- THIS IS THE ORIGINAL FORCE PIXELS IMPLEMENTATION OF super.pointerScroll(delta) --------------------
      // forcePixels(targetPixels);
      // didStartScroll();
      // didUpdateScrollPositionBy(pixels - oldPixels);
      // didEndScroll();
      // goBallistic(0.0);
    }
  }
}
```

**Comment by ReinisSprogis** [comment]
> A bit salty here but why should we bother with a workaround on a thing that should just work? This issue has been live since 2019. Any official update from the flutter team regarding this?

I agree. Mouse scrolling on lists are confusing. Items simply disappear off the screen. If not numbered, it's hard to keep track on what I am looking at.

**Comment by kszczek** [comment]
I've just published the [scroll_animator](https://pub.dev/packages/scroll_animator) package, which aims to be a direct port of Chromium's smooth scrolling algorithm (ported along with related unit tests, just to make sure it works as expected). There are some edge cases which I know about and plan to resolve in the coming days, but there's probably a bunch more of them I don't know about yet, so I wouldn't use it in production just yet, but I would be grateful for testing and bug reports. When the package is stable enough, I'll probably try to refactor the code and try to upstream it.

In regards to the previous attempts at resolving this problem, it turns out we don't have to know if the user is scrolling using a mouse wheel or a trackpad. Chromium developers solved that problem by adjusting the algorithm so it works well with small, frequent scroll deltas as well as large, infrequent scroll deltas. See [this bug report](https://issues.chromium.org/issues/41210665) for more details.

**Comment by BennyKirschner** [comment]
> I've just published the [scroll_animator](https://pub.dev/packages/scroll_animator) package, which aims to be a direct port of Chromium's smooth scrolling algorithm (ported along with related unit tests, just to make sure it works as expected). There are some edge cases which I know about and plan to resolve in the coming days, but there's probably a bunch more of them I don't know about yet, so I wouldn't use it in production just yet, but I would be grateful for testing and bug reports. When the package is stable enough, I'll probably try to refactor the code and try to upstream it.
> 
> In regards to the previous attempts at resolving this problem, it turns out we don't have to know if the user is scrolling using a mouse wheel or a trackpad. Chromium developers solved that problem by adjusting the algorithm so it works well with small, frequent scroll deltas as well as large, infrequent scroll deltas. See [this bug report](https://issues.chromium.org/issues/41210665) for more details.

I just gave it a try but when scrolling up/down quickly, the scroll seems to stop completely every now and again. Once fixed, this packages seems like a great fix to the problem.

**Comment by kszczek** [comment]
> I just gave it a try but when scrolling up/down quickly

Yeah, I'll probably have to get a new mouse, my current one has a broken wheel, when scrolling in one direction it sporadically decides to go the other way, so I wasn't able to verify quick scrolls, I'll look into that.

Also, regarding upstreaming into Flutter, I think it would be best if Flutter provided an abstract animation factory (like the one in my package), so that third-party packages could provide alternative scroll animation implementations. Flutter could include a single default implementation, but then there would be pub packages for Mac-like or Edge-like scrolling. The recently proposed solution with just a boolean would lack that customizability.

**Comment by kszczek** [comment]
@BennyKirschner unfortunately I wasn't able to reproduce your issue, but I've just published a new version with an additional diagnostic example, available here: https://pub.dev/packages/scroll_animator/example. If you could try to reproduce your issue while capturing this diagnostic data, I might be able to synthetically reproduce your scroll inputs and hopefully the issue itself. Also, I suggest opening an issue over at the scroll_animator repository, so we don't spam this issue.

**Comment by kszczek** [comment]
I wanted to share a little progress update, since I've published the initial version of scroll_animator last week, quite a bit changed:

- **Introduced support for programmatic scrolling with automatic animation parameters.**
  
  Using Flutter's [ScrollController.animateTo](https://api.flutter.dev/flutter/widgets/ScrollController/animateTo.html) method requires explicitly providing an animation curve and duration, which can lead to inconsistencies in your app if you use programmatic scroll animations in multiple places.

  Starting with version 0.2.0, using scroll_animator's [AnimatedScrollController.animateTo](https://pub.dev/documentation/scroll_animator/latest/scroll_animator/AnimatedScrollController/animateTo.html) method, you can omit the curve and duration, and they will be determined automagically based on the scroll distance.
  This makes the API similar to the [scrollTo](https://developer.mozilla.org/en-US/docs/Web/API/Window/scrollTo) method you might know from JavaScript. Actually, the scroll animation should feel *exactly* the same as in a Chromium-based browser, pixel-by-pixel.

- **Introduced support for scrolling with arrows and page up/down keys (or any other custom keys)**

  Although Flutter has support for scrolling with the keyboard out-of-the box, it's not really customizable, and is partly broken (try holding down an arrow key to scroll faster). With version 0.3.0, scroll_animator introduces a subclass of Flutter's default scroll action, the [AnimatedScrollAction](https://pub.dev/documentation/scroll_animator/latest/scroll_animator/AnimatedScrollAction-class.html). This action respects the [ScrollAnimationFactory](https://pub.dev/documentation/scroll_animator/latest/scroll_animator/ScrollAnimationFactory-class.html) specified by the user, and also fixes the aforementioned hold-down scroll issue.

- **Improved developer experience**

  Version 0.3.0 of scroll_animator introduces the [AnimatedPrimaryScrollController](https://pub.dev/documentation/scroll_animator/latest/scroll_animator/AnimatedPrimaryScrollController-class.html). This widget creates and manages the [AnimatedScrollController](https://pub.dev/documentation/scroll_animator/latest/scroll_animator/AnimatedScrollController-class.html) instance for you, so you don't have to worry about creating stateful widgets to store the scroll controllers. This also makes the keyboard scrolling work out-of-the box, by wrapping your subtree with a [PrimaryScrollController](https://api.flutter.dev/flutter/widgets/PrimaryScrollController-class.html).

  The example application in the package's README was revamped to exhaustively demonstrate all features of the package.

**Comment by water-mizuu** [comment]
@kszczek, First I have to say, wonderful work with the package. It works flawlessly, and honestly it's better than some of the previous solutions. I have to ask, if it's possible for you to add support for the package `scrollable_positioned_list`? As it uses its own scroll controller class. Cheers!

**Comment by kszczek** [comment]
@water-mizuu thanks for the feedback, I appreciate it!

Regarding the package you've mentioned, I've glanced over its source and it seems like it has the base `ScrollController` type [hardcoded](https://github.com/google/flutter.widgets/blob/494c6abd3de44a92d34c5cbc424db2eefe3915cf/packages/scrollable_positioned_list/lib/src/scrollable_positioned_list.dart#L661), so there is nothing I can do in my package to make this work. To integrate `scroll_animator` with this package, we'd have to be able to provide a custom `ScrollController` instance to the list widget or make it source the scroll controller from a `PrimaryScrollController` ancestor.

I'd submit a PR as it's a quite simple thing to do, but it seems like the [google/flutter.widgets](https://github.com/google/flutter.widgets) repository where this package lives is not very active, with the last commit dating December 2023 and multiple stale PRs. If I were you I'd just clone this package to my machine, edit [this](https://github.com/google/flutter.widgets/blob/494c6abd3de44a92d34c5cbc424db2eefe3915cf/packages/scrollable_positioned_list/lib/src/scrollable_positioned_list.dart#L661) line to instantiate an [AnimatedScrollController](https://pub.dev/documentation/scroll_animator/latest/scroll_animator/AnimatedScrollController-class.html) and use the [path dependency](https://docs.flutter.dev/packages-and-plugins/using-packages#dependencies-on-unpublished-packages) syntax.

**Comment by hruzgar** [comment]
I just tried @kszczek's scroll_animator package and holy crap it works perfectly. It feels like proper native scrolling instead of feeling like some bad mobile port. 
Desktop flutter apps never really felt great to me and now I realize that  scrolling was playing a big part in that. But this just works perfectly and I'm so amazed. Thank you so much for the hard work @kszczek!

**Comment by yjbanov** [comment]
Just cross-linking to a related issue over on the W3C side: https://github.com/w3c/uievents/issues/337

**Comment by hasali19** [comment]
I get why this is currently blocked for web, but is there anything preventing this from being implemented for desktop platforms?

**Comment by kszczek** [comment]
> I get why this is currently blocked for web

I don't think differentiating between trackpad and mouse wheel events should be considered a blocker for web/desktop implementation. In fact, I believe the reason stated for this blockage is a misunderstanding.

I've mentioned this previously, but it's worth reiterating: some years ago, Chromium developers also implemented a heuristic for differentiating between trackpad and mouse wheel inputs. However, this approach caused more issues than it resolved. So, instead of disabling scroll smoothing specifically for trackpads, Chromium developers improved the scroll smoothing algorithm to handle high-frequency, low-delta inputs better (see https://issues.chromium.org/issues/41210665#comment16).

All of the above is highlighted in the `scroll_animator`'s readme:

> Trackpads typically provide precise and frequent scroll deltas, which often leads to the assumption that scroll smoothing isn't necessary. However, some trackpads can produce larger, less frequent deltas, similar to mouse wheels. Therefore, this package handles all scroll inputs without making assumptions based on the pointer type.
>
> See [this Chromium bug](https://issues.chromium.org/41210665) for discussion on this topic.

As described above, `scroll_animator` handles trackpad and mouse wheel inputs identically, and yet scrolling behaves as expected (at least on Linux and Windows - I haven't tested this on a MacBook's trackpad yet, so feedback on its behavior would be greatly appreciated!).

I was in the process of upstreaming my work into Flutter last year, but got overwhelmed with other work. I plan to get back to this soon though, here's a rough roadmap:

1. Design an API for implementing smooth scrolling algorithms (done - https://flutter.dev/go/smoothing-discrete-scroll-inputs)
2. Refactor some existing Flutter components to make them compatible with the new API, as outlined in the doc.
3. Implement and expose the bare API, without any actual implementations.
4. Convert `scroll_animator` package into a "catalogue" of scroll smoothing algorithms - initially including Chrome and MS Edge algorithms (IIRC Windows UWP scroll smoothing algo is open source, so should be easy to port also).
5. Implement a platform-agnostic scroll smoothing algorithm and include it in the framework(?)

**Comment by kszczek** [comment]
(although knowing that a trackpad is used for scrolling is still useful info. It would allow us to implement fling gestures. But its still not a blocker - only a cherry on top)

**Comment by JHubi1** [comment]
It's really bad right now, and I don't get why this was downgraded from P1 to P2. It definitely deserves highest priority and should be considered a severe bug.
There's no real fix available that does not make the UX significantly worse in other spots. All proposed seem to break something or are a pain to implement consistently.
An issue this bad really stops devs from using Flutter in any serious project that's supposed to run on any of the Desktop platforms. Really needs a fix

**Comment by shroot91** [comment]
Why this has been marked like a Feature ? It's a bug.

**Comment by iapicca** [comment]
I don't think @shroot91 comment is spam
it is exactly as he stated, labeling this issue as "new feature"
while from developer side is clearly a bug
misrepresents the urgency of the issue

**Comment by Piinks** [comment]
Happy to clarify! This is a bug, that requires a new feature in order to resolve. 

This bug in particular has also been plagued with spam, in spite of our documented [issue hygiene](https://github.com/flutter/flutter/blob/master/docs/contributing/issue_hygiene/README.md), so yes, folks that are subscribed for meaningful updates are moderating it heavily.

**Comment by Krushiler** [comment]
[iapicca](https://github.com/iapicca) Now it looks like it's better not to target for web and Windows. Every app has scrollable components but it's very unconfortable to scroll with mouse and this issue still remains open. After 6 years

**Comment by flutter-zl** [comment]
> > I get why this is currently blocked for web
> 
> I don't think differentiating between trackpad and mouse wheel events should be considered a blocker for web/desktop implementation. In fact, I believe the reason stated for this blockage is a misunderstanding.
> 
> I've mentioned this previously, but it's worth reiterating: some years ago, Chromium developers also implemented a heuristic for differentiating between trackpad and mouse wheel inputs. However, this approach caused more issues than it resolved. So, instead of disabling scroll smoothing specifically for trackpads, Chromium developers improved the scroll smoothing algorithm to handle high-frequency, low-delta inputs better (see https://issues.chromium.org/issues/41210665#comment16).
> 
> All of the above is highlighted in the `scroll_animator`'s readme:
> 
> > Trackpads typically provide precise and frequent scroll deltas, which often leads to the assumption that scroll smoothing isn't necessary. However, some trackpads can produce larger, less frequent deltas, similar to mouse wheels. Therefore, this package handles all scroll inputs without making assumptions based on the pointer type.
> > See [this Chromium bug](https://issues.chromium.org/41210665) for discussion on this topic.
> 
> As described above, `scroll_animator` handles trackpad and mouse wheel inputs identically, and yet scrolling behaves as expected (at least on Linux and Windows - I haven't tested this on a MacBook's trackpad yet, so feedback on its behavior would be greatly appreciated!).
> 
> I was in the process of upstreaming my work into Flutter last year, but got overwhelmed with other work. I plan to get back to this soon though, here's a rough roadmap:
> 
> 1. Design an API for implementing smooth scrolling algorithms (done - https://flutter.dev/go/smoothing-discrete-scroll-inputs)
> 2. Refactor some existing Flutter components to make them compatible with the new API, as outlined in the doc.
> 3. Implement and expose the bare API, without any actual implementations.
> 4. Convert `scroll_animator` package into a "catalogue" of scroll smoothing algorithms - initially including Chrome and MS Edge algorithms (IIRC Windows UWP scroll smoothing algo is open source, so should be easy to port also).
> 5. Implement a platform-agnostic scroll smoothing algorithm and include it in the framework(?)

Hi Kamil,
Your [scroll_animator package](https://pub.dev/packages/scroll_animator) works incredibly well on Flutter We. The smooth scrolling experience is a huge improvement. Would you be interested in contributing this functionality directly to the Flutter framework? I think it could bring a lot of value to the wider community. Let me know if you’d be open to collaborating or exploring how to make that happen.

**Comment by kszczek** [comment]
@flutter-zl

> Would you be interested in contributing this functionality directly to the Flutter framework?

I'd love to!

Looking back on my notes from a few months back, I can see that the main roadblock for landing this was the NestedScrollView, which would have to accept absolute scroll offsets in order to work with the planned Smooth Scrolling API (specifically it would have to implement [setPixels](https://github.com/flutter/flutter/blob/22613b764df436ce57ead1f0fe56cebe6a9238c0/packages/flutter/lib/src/widgets/nested_scroll_view.dart#L1023), see my [design doc](https://docs.google.com/document/d/136wxl8He3PF2UMOT51lPfeqKE2TWGQuQ1u1b-ZDgErM/edit?tab=t.0#heading=h.vuf4ow52ej6k) for more details).

I was going to refactor the NestedScrollView to make it compatible, but then I stumbled upon issue #102003, which outlines plans to rewrite it completely, so I decided it was not worth putting any significant effort into the current implementation, which would eventually be wiped anyway.

We _could_ however decide to ignore the NestedScrollView for now, and wire it up with the Smooth Scrolling API as part of the rewrite. In that case, I already have a POC implementation from a few months back which I can polish up a bit and put a PR up for.

**Comment by flutter-zl** [comment]
> [@flutter-zl](https://github.com/flutter-zl)
> 
> > Would you be interested in contributing this functionality directly to the Flutter framework?
> 
> I'd love to!
> 
> Looking back on my notes from a few months back, I can see that the main roadblock for landing this was the NestedScrollView, which would have to accept absolute scroll offsets in order to work with the planned Smooth Scrolling API (specifically it would have to implement [setPixels](https://github.com/flutter/flutter/blob/22613b764df436ce57ead1f0fe56cebe6a9238c0/packages/flutter/lib/src/widgets/nested_scroll_view.dart#L1023), see my [design doc](https://docs.google.com/document/d/136wxl8He3PF2UMOT51lPfeqKE2TWGQuQ1u1b-ZDgErM/edit?tab=t.0#heading=h.vuf4ow52ej6k) for more details).
> 
> I was going to refactor the NestedScrollView to make it compatible, but then I stumbled upon issue [#102003](https://github.com/flutter/flutter/issues/102003), which outlines plans to rewrite it completely, so I decided it was not worth putting any significant effort into the current implementation, which would eventually be wiped anyway.
> 
> We _could_ however decide to ignore the NestedScrollView for now, and wire it up with the Smooth Scrolling API as part of the rewrite. In that case, I already have a POC implementation from a few months back which I can polish up a bit and put a PR up for.

I’m really glad you’re open to contributing! If you already have a working POC, polishing it into a PR would be a fantastic starting point. I’d be happy to help review or coordinate with the Flutter framework team to get things moving.

**Comment by Piinks** [comment]
Thanks @kszczek! Happy to support as well!

> I was going to refactor the NestedScrollView to make it compatible, but then I stumbled upon issue https://github.com/flutter/flutter/issues/102003, which outlines plans to rewrite it completely, so I decided it was not worth putting any significant effort into the current implementation, which would eventually be wiped anyway.

No one is working on rewriting the NestedScrollView currently, and I don't think anyone is planning to, so no worries about it being blown away. :)

> the main roadblock for landing this was the NestedScrollView, which would have to accept absolute scroll offsets in order to work with the planned Smooth Scrolling API (specifically it would have to implement [setPixels](https://github.com/flutter/flutter/blob/22613b764df436ce57ead1f0fe56cebe6a9238c0/packages/flutter/lib/src/widgets/nested_scroll_view.dart#L1023), see my [design doc](https://docs.google.com/document/d/136wxl8He3PF2UMOT51lPfeqKE2TWGQuQ1u1b-ZDgErM/edit?tab=t.0#heading=h.vuf4ow52ej6k) for more details).

I'd love to discuss this further and see fi we can make this change in NestedScrollView, let's discuss this further, I'd love to understand what would be required.

**Comment by clragon** [comment]
> @flutter-zl said:
> [entirely AI generated reply, as indicated by both the prominent use of emdash, as well as the sentence structure]

Without wanting to derail this extremely important topic at hand, which I have been invested in since years, because I would like to eventually release my app on desktop;
Please do afford other people the human decency of writing your replies yourself. It is harrowing to see important technical discussions degrade like this.

**Comment by JHubi1** [comment]
> [@flutter-zl](https://github.com/flutter-zl) said:
> [entirely AI generated reply, as indicated by both the prominent use of emdash, as well as the sentence structure]

I don’t see an issue with using the correct English typography in a piece of writing. Simply pointing out a few set characters in a text—like an em dash—does not accurately reflect a person’s realness. The text was also way too short to actually be relevant sentence structure wise.
Instead of trying to “hunt down” alleged AI‐written texts, you could instead try to actually contribute positively to the situation. Find a way to contribute to the situation, or be a silent watcher.
And yes, I used a provocative amount of typographically correct characters in this text.

**Comment by luoyi58624** [comment]
https://luoyi58624.github.io/smooth_scroll_controller
Simply implemented a smooth scrolling effect(only 200 line code), It runs quite well on my Windows device, Perhaps it would be helpful to refer to it? https://github.com/luoyi58624/smooth_scroll_controller
